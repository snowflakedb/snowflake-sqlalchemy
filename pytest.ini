# pytest.ini
[pytest]
# There are majorly two issues with the sqlalchemy test suites if we want to run it with other snowflake specific tests:
# 1. If including the test_suite.py, only tests in test_suite.py will be collected, other dialect tests are not collected
# due to "from sqlalchemy.testing.plugin.pytestplugin import".
# 2. Putting test_suites.py and related conftest config under tests root dir would make pytest fail to collect cfg
# information such as requirements, and customized session db config does not work as expected (schema not set).
# Thus, we ignore the path for now, and give instructions to run the test suites in standalone directory.
addopts = --ignore=tests/sqlalchemy_test_suite
