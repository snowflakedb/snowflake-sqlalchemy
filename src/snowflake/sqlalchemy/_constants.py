#
# Copyright (c) 2012-2022 Snowflake Computing Inc. All rights reserved.
#
import pkg_resources

# parameters needed for usage tracking
PARAM_APPLICATION = "application"
PARAM_INTERNAL_APPLICATION_NAME = "internal_application_name"
PARAM_INTERNAL_APPLICATION_VERSION = "internal_application_version"

APPLICATION_NAME = "SnowflakeSQLAlchemy"
SNOWFLAKE_SQLALCHEMY_VERSION = pkg_resources.get_distribution(
    "snowflake-sqlalchemy"
).version
