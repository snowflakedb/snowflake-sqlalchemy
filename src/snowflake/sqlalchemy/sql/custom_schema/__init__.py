#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
#
from .dynamic_table import DynamicTable
from .hybrid_table import HybridTable

__all__ = ["DynamicTable", "HybridTable"]
