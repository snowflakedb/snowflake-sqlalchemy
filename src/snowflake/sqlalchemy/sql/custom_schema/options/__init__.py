#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
#

from .as_query import AsQuery
from .target_lag import TargetLag, TimeUnit
from .warehouse import Warehouse

__all__ = ["Warehouse", "AsQuery", "TargetLag", "TimeUnit"]
