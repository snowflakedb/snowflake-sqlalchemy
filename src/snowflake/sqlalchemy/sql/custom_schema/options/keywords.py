#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.

from enum import Enum


class SnowflakeKeyword(Enum):
    # TARGET_LAG
    DOWNSTREAM = "DOWNSTREAM"

    # REFRESH_MODE
    AUTO = "AUTO"
    FULL = "FULL"
    INCREMENTAL = "INCREMENTAL"
