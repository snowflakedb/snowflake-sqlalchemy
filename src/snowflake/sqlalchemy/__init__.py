#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
#

import sys

if sys.version_info < (3, 8):
    import importlib_metadata
else:
    import importlib.metadata as importlib_metadata

from sqlalchemy.types import (  # noqa
    BIGINT,
    BINARY,
    BOOLEAN,
    CHAR,
    DATE,
    DATETIME,
    DECIMAL,
    FLOAT,
    INT,
    INTEGER,
    REAL,
    SMALLINT,
    TIME,
    TIMESTAMP,
    VARCHAR,
)

from . import base, snowdialect  # noqa
from .custom_commands import (  # noqa
    AWSBucket,
    AzureContainer,
    CopyFormatter,
    CopyIntoStorage,
    CreateFileFormat,
    CreateStage,
    CSVFormatter,
    ExternalStage,
    JSONFormatter,
    MergeInto,
    PARQUETFormatter,
)
from .custom_types import (  # noqa
    ARRAY,
    BYTEINT,
    CHARACTER,
    DEC,
    DOUBLE,
    FIXED,
    GEOGRAPHY,
    GEOMETRY,
    MAP,
    NUMBER,
    OBJECT,
    STRING,
    TEXT,
    TIMESTAMP_LTZ,
    TIMESTAMP_NTZ,
    TIMESTAMP_TZ,
    TINYINT,
    VARBINARY,
    VARIANT,
)
from .sql.custom_schema import (  # noqa
    DynamicTable,
    HybridTable,
    IcebergTable,
    SnowflakeTable,
)
from .sql.custom_schema.options import (  # noqa
    AsQueryOption,
    ClusterByOption,
    IdentifierOption,
    KeywordOption,
    LiteralOption,
    SnowflakeKeyword,
    TableOptionKey,
    TargetLagOption,
    TimeUnit,
)
from .util import _url as URL  # noqa

base.dialect = dialect = snowdialect.dialect

__version__ = importlib_metadata.version("snowflake-sqlalchemy")

_custom_types = (
    "BIGINT",
    "BINARY",
    "BOOLEAN",
    "CHAR",
    "DATE",
    "DATETIME",
    "DECIMAL",
    "FLOAT",
    "INT",
    "INTEGER",
    "REAL",
    "SMALLINT",
    "TIME",
    "TIMESTAMP",
    "URL",
    "VARCHAR",
    "ARRAY",
    "BYTEINT",
    "CHARACTER",
    "DEC",
    "DOUBLE",
    "FIXED",
    "GEOGRAPHY",
    "GEOMETRY",
    "OBJECT",
    "NUMBER",
    "STRING",
    "TEXT",
    "TIMESTAMP_LTZ",
    "TIMESTAMP_TZ",
    "TIMESTAMP_NTZ",
    "TINYINT",
    "VARBINARY",
    "VARIANT",
    "MAP",
)

_custom_commands = (
    "MergeInto",
    "CSVFormatter",
    "JSONFormatter",
    "PARQUETFormatter",
    "CopyFormatter",
    "CopyIntoStorage",
    "AWSBucket",
    "AzureContainer",
    "ExternalStage",
    "CreateStage",
    "CreateFileFormat",
)

_custom_tables = ("HybridTable", "DynamicTable", "IcebergTable", "SnowflakeTable")

_custom_table_options = (
    "AsQueryOption",
    "TargetLagOption",
    "LiteralOption",
    "IdentifierOption",
    "KeywordOption",
    "ClusterByOption",
)

_enums = (
    "TimeUnit",
    "TableOptionKey",
    "SnowflakeKeyword",
)
__all__ = (
    *_custom_types,
    *_custom_commands,
    *_custom_tables,
    *_custom_table_options,
    *_enums,
)
