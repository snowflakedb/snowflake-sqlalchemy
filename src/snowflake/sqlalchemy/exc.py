#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.

from typing import List

from sqlalchemy.exc import ArgumentError


class NoPrimaryKeyError(ArgumentError):
    def __init__(self, target: str):
        super().__init__(f"Table {target} required primary key.")


class UnsupportedPrimaryKeysAndForeignKeysError(ArgumentError):
    def __init__(self, target: str):
        super().__init__(f"Primary key and foreign keys are not supported in {target}.")


class RequiredParametersNotProvidedError(ArgumentError):
    def __init__(self, target: str, parameters: List[str]):
        super().__init__(
            f"{target} requires the following parameters: %s." % ", ".join(parameters)
        )


class UnexpectedTableOptionKeyError(ArgumentError):
    def __init__(self, expected: str, actual: str):
        super().__init__(f"Expected table option {expected} but got {actual}.")


class OptionKeyNotProvidedError(ArgumentError):
    def __init__(self, target: str):
        super().__init__(
            f"Expected option key in {target} option but got NoneType instead."
        )


class UnexpectedOptionParameterTypeError(ArgumentError):
    def __init__(self, parameter_name: str, target: str, types: List[str]):
        super().__init__(
            f"Parameter {parameter_name} of {target} requires to be one"
            f" of following types: {', '.join(types)}."
        )


class CustomOptionsAreOnlySupportedOnSnowflakeTables(ArgumentError):
    def __init__(self):
        super().__init__(
            "Identifier, Literal, TargetLag and other custom options are only supported on Snowflake tables."
        )


class UnexpectedOptionTypeError(ArgumentError):
    def __init__(self, options: List[str]):
        super().__init__(
            f"The following options are either unsupported or should be defined using a Snowflake table: {', '.join(options)}."
        )


class InvalidTableParameterTypeError(ArgumentError):
    def __init__(self, name: str, input_type: str, expected_types: List[str]):
        expected_types_str = "', '".join(expected_types)
        super().__init__(
            f"Invalid parameter type '{input_type}' provided for '{name}'. "
            f"Expected one of the following types: '{expected_types_str}'.\n"
        )


class MultipleErrors(ArgumentError):
    def __init__(self, errors):
        self.errors = errors

    def __str__(self):
        return "".join(str(e) for e in self.errors)


class StructuredTypeNotSupportedInTableColumnsError(ArgumentError):
    def __init__(self, table_type: str, table_name: str, column_name: str):
        super().__init__(
            f"Column '{column_name}' is of a structured type, which is only supported on Iceberg tables. "
            f"The table '{table_name}' is of type '{table_type}', not Iceberg."
        )
