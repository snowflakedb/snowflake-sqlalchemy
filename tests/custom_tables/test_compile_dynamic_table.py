#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
#
import pytest
from sqlalchemy import (
    Column,
    ForeignKeyConstraint,
    Integer,
    MetaData,
    String,
    Table,
    exc,
    select,
)
from sqlalchemy.exc import ArgumentError
from sqlalchemy.orm import declarative_base
from sqlalchemy.sql.ddl import CreateTable

from snowflake.sqlalchemy import GEOMETRY, DynamicTable
from snowflake.sqlalchemy.exc import MultipleErrors
from snowflake.sqlalchemy.sql.custom_schema.options import (
    AsQueryOption,
    IdentifierOption,
    KeywordOption,
    LiteralOption,
    TargetLagOption,
    TimeUnit,
)
from snowflake.sqlalchemy.sql.custom_schema.options.keywords import SnowflakeKeyword


def test_compile_dynamic_table(sql_compiler, snapshot):
    metadata = MetaData()
    table_name = "test_dynamic_table"
    test_geometry = DynamicTable(
        table_name,
        metadata,
        Column("id", Integer),
        Column("geom", GEOMETRY),
        target_lag=(10, TimeUnit.SECONDS),
        warehouse="warehouse",
        as_query="SELECT * FROM table",
    )

    value = CreateTable(test_geometry)

    actual = sql_compiler(value)

    assert actual == snapshot


@pytest.mark.parametrize(
    "refresh_mode_keyword",
    [
        SnowflakeKeyword.AUTO,
        SnowflakeKeyword.FULL,
        SnowflakeKeyword.INCREMENTAL,
    ],
)
def test_compile_dynamic_table_with_refresh_mode(
    sql_compiler, snapshot, refresh_mode_keyword
):
    metadata = MetaData()
    table_name = "test_dynamic_table"
    test_geometry = DynamicTable(
        table_name,
        metadata,
        Column("id", Integer),
        Column("geom", GEOMETRY),
        target_lag=(10, TimeUnit.SECONDS),
        warehouse="warehouse",
        as_query="SELECT * FROM table",
        refresh_mode=refresh_mode_keyword,
    )

    value = CreateTable(test_geometry)

    actual = sql_compiler(value)

    assert actual == snapshot


def test_compile_dynamic_table_with_options_objects(sql_compiler, snapshot):
    metadata = MetaData()
    table_name = "test_dynamic_table"
    test_geometry = DynamicTable(
        table_name,
        metadata,
        Column("id", Integer),
        Column("geom", GEOMETRY),
        target_lag=TargetLagOption(10, TimeUnit.SECONDS),
        warehouse=IdentifierOption("warehouse"),
        as_query=AsQueryOption("SELECT * FROM table"),
        refresh_mode=KeywordOption(SnowflakeKeyword.AUTO),
    )

    value = CreateTable(test_geometry)

    actual = sql_compiler(value)

    assert actual == snapshot


def test_compile_dynamic_table_with_one_wrong_option_types(snapshot):
    metadata = MetaData()
    table_name = "test_dynamic_table"
    with pytest.raises(ArgumentError) as argument_error:
        DynamicTable(
            table_name,
            metadata,
            Column("id", Integer),
            Column("geom", GEOMETRY),
            target_lag=TargetLagOption(10, TimeUnit.SECONDS),
            warehouse=LiteralOption("warehouse"),
            as_query=AsQueryOption("SELECT * FROM table"),
            refresh_mode=KeywordOption(SnowflakeKeyword.AUTO),
        )

    assert str(argument_error.value) == snapshot


def test_compile_dynamic_table_with_multiple_wrong_option_types(snapshot):
    metadata = MetaData()
    table_name = "test_dynamic_table"
    with pytest.raises(MultipleErrors) as argument_error:
        DynamicTable(
            table_name,
            metadata,
            Column("id", Integer),
            Column("geom", GEOMETRY),
            target_lag=IdentifierOption(SnowflakeKeyword.AUTO),
            warehouse=KeywordOption(SnowflakeKeyword.AUTO),
            as_query=KeywordOption(SnowflakeKeyword.AUTO),
            refresh_mode=IdentifierOption(SnowflakeKeyword.AUTO),
        )

    assert str(argument_error.value) == snapshot


def test_compile_dynamic_table_without_required_args(sql_compiler):
    with pytest.raises(
        exc.ArgumentError,
        match="DynamicTable requires the following parameters: warehouse, "
        "as_query, target_lag.",
    ):
        DynamicTable(
            "test_dynamic_table",
            MetaData(),
            Column("id", Integer, primary_key=True),
            Column("geom", GEOMETRY),
        )


def test_compile_dynamic_table_with_primary_key(sql_compiler):
    with pytest.raises(
        exc.ArgumentError,
        match="Primary key and foreign keys are not supported in DynamicTable.",
    ):
        DynamicTable(
            "test_dynamic_table",
            MetaData(),
            Column("id", Integer, primary_key=True),
            Column("geom", GEOMETRY),
            target_lag=(10, TimeUnit.SECONDS),
            warehouse="warehouse",
            as_query="SELECT * FROM table",
        )


def test_compile_dynamic_table_with_foreign_key(sql_compiler):
    with pytest.raises(
        exc.ArgumentError,
        match="Primary key and foreign keys are not supported in DynamicTable.",
    ):
        DynamicTable(
            "test_dynamic_table",
            MetaData(),
            Column("id", Integer),
            Column("geom", GEOMETRY),
            ForeignKeyConstraint(["id"], ["table.id"]),
            target_lag=(10, TimeUnit.SECONDS),
            warehouse="warehouse",
            as_query="SELECT * FROM table",
        )


def test_compile_dynamic_table_orm(sql_compiler, snapshot):
    Base = declarative_base()
    metadata = MetaData()
    table_name = "test_dynamic_table_orm"
    test_dynamic_table_orm = DynamicTable(
        table_name,
        metadata,
        Column("id", Integer),
        Column("name", String),
        target_lag=(10, TimeUnit.SECONDS),
        warehouse="warehouse",
        as_query="SELECT * FROM table",
    )

    class TestDynamicTableOrm(Base):
        __table__ = test_dynamic_table_orm
        __mapper_args__ = {
            "primary_key": [test_dynamic_table_orm.c.id, test_dynamic_table_orm.c.name]
        }

        def __repr__(self):
            return f"<TestDynamicTableOrm({self.name!r}, {self.fullname!r})>"

    value = CreateTable(TestDynamicTableOrm.__table__)

    actual = sql_compiler(value)

    assert actual == snapshot


def test_compile_dynamic_table_orm_with_str_keys(sql_compiler, snapshot):
    Base = declarative_base()

    class TestDynamicTableOrm(Base):
        __tablename__ = "test_dynamic_table_orm_2"

        @classmethod
        def __table_cls__(cls, name, metadata, *arg, **kw):
            return DynamicTable(name, metadata, *arg, **kw)

        __table_args__ = {
            "schema": "SCHEMA_DB",
            "target_lag": (10, TimeUnit.SECONDS),
            "warehouse": "warehouse",
            "as_query": "SELECT * FROM table",
        }

        id = Column(Integer)
        name = Column(String)

        __mapper_args__ = {"primary_key": [id, name]}

        def __repr__(self):
            return f"<TestDynamicTableOrm({self.name!r}, {self.fullname!r})>"

    value = CreateTable(TestDynamicTableOrm.__table__)

    actual = sql_compiler(value)

    assert actual == snapshot


def test_compile_dynamic_table_with_selectable(sql_compiler, snapshot):
    Base = declarative_base()

    test_table_1 = Table(
        "test_table_1",
        Base.metadata,
        Column("id", Integer, primary_key=True),
        Column("name", String),
    )

    dynamic_test_table = DynamicTable(
        "dynamic_test_table_1",
        Base.metadata,
        target_lag=(10, TimeUnit.SECONDS),
        warehouse="warehouse",
        as_query=select(test_table_1).where(test_table_1.c.id == 23),
    )

    value = CreateTable(dynamic_test_table)

    actual = sql_compiler(value)

    assert actual == snapshot
