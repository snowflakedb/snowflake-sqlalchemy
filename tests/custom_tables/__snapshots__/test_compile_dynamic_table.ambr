# serializer version: 1
# name: test_compile_dynamic_table
  "CREATE DYNAMIC TABLE test_dynamic_table (\tid INTEGER, \tgeom GEOMETRY)\tWAREHOUSE = warehouse\tTARGET_LAG = '10 seconds'\tAS SELECT * FROM table"
# ---
# name: test_compile_dynamic_table_orm
  "CREATE DYNAMIC TABLE test_dynamic_table_orm (\tid INTEGER, \tname VARCHAR)\tWAREHOUSE = warehouse\tTARGET_LAG = '10 seconds'\tAS SELECT * FROM table"
# ---
# name: test_compile_dynamic_table_orm_with_str_keys
  'CREATE DYNAMIC TABLE "SCHEMA_DB".test_dynamic_table_orm_2 (\tid INTEGER, \tname VARCHAR)\tWAREHOUSE = warehouse\tTARGET_LAG = \'10 seconds\'\tAS SELECT * FROM table'
# ---
# name: test_compile_dynamic_table_with_multiple_wrong_option_types
  '''
  Invalid parameter type 'IdentifierOption' provided for 'refresh_mode'. Expected one of the following types: 'KeywordOption', 'SnowflakeKeyword'.
  Invalid parameter type 'IdentifierOption' provided for 'target_lag'. Expected one of the following types: 'TargetLagOption', 'Tuple[int, TimeUnit])', 'SnowflakeKeyword'.
  Invalid parameter type 'KeywordOption' provided for 'as_query'. Expected one of the following types: 'AsQueryOption', 'str', 'Selectable'.
  Invalid parameter type 'KeywordOption' provided for 'warehouse'. Expected one of the following types: 'IdentifierOption', 'str'.
  
  '''
# ---
# name: test_compile_dynamic_table_with_one_wrong_option_types
  '''
  Invalid parameter type 'LiteralOption' provided for 'warehouse'. Expected one of the following types: 'IdentifierOption', 'str'.
  
  '''
# ---
# name: test_compile_dynamic_table_with_options_objects
  "CREATE DYNAMIC TABLE test_dynamic_table (\tid INTEGER, \tgeom GEOMETRY)\tWAREHOUSE = warehouse\tTARGET_LAG = '10 seconds'\tREFRESH_MODE = AUTO\tAS SELECT * FROM table"
# ---
# name: test_compile_dynamic_table_with_refresh_mode[SnowflakeKeyword.AUTO]
  "CREATE DYNAMIC TABLE test_dynamic_table (\tid INTEGER, \tgeom GEOMETRY)\tWAREHOUSE = warehouse\tTARGET_LAG = '10 seconds'\tREFRESH_MODE = AUTO\tAS SELECT * FROM table"
# ---
# name: test_compile_dynamic_table_with_refresh_mode[SnowflakeKeyword.FULL]
  "CREATE DYNAMIC TABLE test_dynamic_table (\tid INTEGER, \tgeom GEOMETRY)\tWAREHOUSE = warehouse\tTARGET_LAG = '10 seconds'\tREFRESH_MODE = FULL\tAS SELECT * FROM table"
# ---
# name: test_compile_dynamic_table_with_refresh_mode[SnowflakeKeyword.INCREMENTAL]
  "CREATE DYNAMIC TABLE test_dynamic_table (\tid INTEGER, \tgeom GEOMETRY)\tWAREHOUSE = warehouse\tTARGET_LAG = '10 seconds'\tREFRESH_MODE = INCREMENTAL\tAS SELECT * FROM table"
# ---
# name: test_compile_dynamic_table_with_selectable
  "CREATE DYNAMIC TABLE dynamic_test_table_1 (\tid INTEGER, \tname VARCHAR)\tWAREHOUSE = warehouse\tTARGET_LAG = '10 seconds'\tAS SELECT test_table_1.id, test_table_1.name FROM test_table_1 WHERE test_table_1.id = 23"
# ---
